pkg/util/scheduler/scheduler.go

```diff
@@ -0,0 +1,203 @@
+package scheduler
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/grafana/dskit/backoff"
+	"github.com/grafana/dskit/services"
+	"github.com/grafana/grafana/pkg/infra/log"
+)
+
+const (
+	// DefaultMaxBackoff is the default maximum backoff duration for workers.
+	DefaultMaxBackoff = 1 * time.Second
+	// DefaultMinBackoff is the default minimum backoff duration for workers.
+	DefaultMinBackoff = 100 * time.Millisecond
+	// DefaultNumWorkers is the default number of workers in the scheduler.
+	DefaultNumWorkers = 4
+	// DefaultMaxRetries is the default maximum number of retries for dequeue operations.
+	DefaultMaxRetries = 5
+)
+
+type WorkQueue interface {
+	services.Service
+
+	Dequeue(ctx context.Context) (runnable func(), err error)
+}
+
+// Worker processes items from the QoS request queue
+type Worker struct {
+	id         int
+	queue      WorkQueue
+	wg         *sync.WaitGroup
+	maxBackoff time.Duration
+	maxRetries int
+	logger     log.Logger
+}
+
+func (w *Worker) run(ctx context.Context) {
+	defer w.wg.Done()
+	w.logger.Debug("worker started", "id", w.id)
+
+	for {
+		select {
+		case <-ctx.Done():
+			w.logger.Debug("worker stopped", "id", w.id)
+			return
+		default:
+			w.dequeueWithRetries(ctx)
+		}
+	}
+}
+
+func (w *Worker) dequeueWithRetries(ctx context.Context) {
+	boff := backoff.New(ctx, backoff.Config{
+		MinBackoff: DefaultMinBackoff,
+		MaxBackoff: w.maxBackoff,
+		MaxRetries: w.maxRetries,
+	})
+
+	for boff.Ongoing() {
+		runnable, err := w.queue.Dequeue(ctx)
+		if err == nil {
+			runnable()
```

Comment:
Should we pass context to the executed runnables? If worker is stopping, how does it signal runnable that it should stop too?